GitHub Actions CI (#66)
andycroll/andycroll.com: _posts/2019/2019-11-03-github-actions-ci-for-rails-with-postgresql.md

---
title: "Use GitHub Actions for Rails CI with Postgres"
description: "So many commits to get here"
layout: article
category: ruby
image:
  base: '2019/github-actions-ci-for-rails-with-postgresql'
  alt: 'Green Octocat testing'

---

> Writing tests when writing software is strongly recommended, it helps you protect against bugs and inadvertent changes, reduce the fear of scary deploys and can even help you make better software in

Reviewer: ```suggestion
Writing tests alongside writing software is strongly recommended. It helps you protect against bugs, reduce the fear of scary deploys, and can even help you write better software.
```

The Ruby and Rails community have always had a strong focus on testing. Typically that means having some sort of “continuous integration” system constantly running your test suite.

> There’s a bunch of good choices, but recently GitHub launched [Actions](https://github.com/features/actions) (in beta) which lets you run arbitrary workflows, including tests, when certain things happen to your git repository.

Reviewer: Comma before "which"

Reviewer: *after certain things happen...

> It does need a little extra configuration to run your tests using PostgreSQL (my SQL database of choice).

Reviewer: Worth saying here? Or will this eventually be the link to the other one? Either way, I think we should remove this from here or re-word it.

## Instead of...

...using a separate CI service.

## Use...

...Github Actions to run your Rails tests.

### `.github/workflows/tests.yml`

```yml
name: Rails Tests

on:
  pull_request:
    - 'master'
  push:
    - 'master'

jobs:
  build:

    runs-on: ubuntu-latest

    services:
      postgres:
        image: postgres:11.5
        ports: ["5432:5432"]
        options: --health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 5

    steps:
    - uses: actions/checkout@v1

    - name: Set up Ruby 2.6
      uses: actions/setup-ruby@v1
      with:
        ruby-version: 2.6.x

    - name: Install PostgreSQL 11 client
      run: |
        sudo apt-get -yqq install libpq-dev

    - name: Build App
      env:
        PGHOST: localhost
        PGUSER: postgres
        RAILS_ENV: test
      run: |
        gem install bundler
        bundle install --jobs 4 --retry 3
        bin/rails db:setup

    - name: Run Tests
      env:
        PGHOST: localhost
        PGUSER: postgres
        RAILS_ENV: test
      run: |
      bundle exec rake test
      bundle exec rake test:system
      # or Rspec
      # bundle exec rspec
```

> If you use SQL dumps—you have a `db/structure.sql` file rather than a `db/schema.rb` file in your application—you’ll need to use [this slightly altered workflow](github-actions-ci-for-rails-with-postgresql-11-and-structure-sql).

Reviewer: Ahhh — then do you need the sentence on line 18?

## Why?

> There is a benefit to your tests being in the ‘same place’ as your code. There is no separate account to maintain, no separate service to bill, no seperate ‘team member management’.

Reviewer: *and no separate 'team member...

> For limited usage it is _free_, even for private repositories, and pretty economical after that. And completely free for public, open source, repositories.

Reviewer: *It's completely free for public...

Although you already say "even for private repos" so maybe talk about free for private first, then go into public.

> Github Actions can also be configured to do many tasks other than just running your tests. Automatic image compression, other code checks, notifications to other systems, adding comments, making assignments, doing deployments and more. All you have to do is fight with the configuration. :-)

Reviewer: Colon after 'running your tests', then lower-case 'a' on ''Automatic".

Comma before 'and more'.

## Why not?

> Your testing infrastructure should, as much as possible, reflect your production environment to help catch subtle bugs. For me, at [work](https://coveragebook.com), that means using Heroku’s [pipelines, CI and review apps](https://www.heroku.com/flow), meaning our testing and QA infrastructure is as similar as we can make it to production.

Reviewer: Don't think you need the last clause here. After all, you already say: your testing infrastructure should reflect prod as much as poss — so the last clause is already implied.

> You might also feel the idea of further centralisation of your development infrastructure with one company to be a risk. Or a different CI provider might offer you higher levels of service or configurability if you have specific needs.

Reviewer: ```suggestion
You might also think that increased centralisation of your development infrastructure at one company is a risk, or that a different CI provider might offer you with higher levels of service or configurability, if you have specific needs.
```

> If you have a testing infrastructure already set up somewhere else, potentially wound into a working system for deployment, it probably isn’t worth the effort to move.

Reviewer: ```suggestion
If you have a testing infrastructure that's working well already set up somewhere else, it probably isn’t worth the effort to move.
```

> There’s also a current issue with staying on the latest version of Ruby.

Reviewer: Huh? This last sentence is not clear at all. And it seems randomly tacked on.

What are you trying to say and how does GitHub Actions hinder this? Unlike Heroku buildpacks, it won't update Ruby for you?

  push:
    - 'master'

jobs:
  build:
    runs-on: ubuntu-latest

    services:
      postgres:
        image: postgres:11.5
        ports: ["5432:5432"]
        options: --health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 5

    steps:
    - uses: actions/checkout@v1

    - name: Set up Ruby 2.6
      uses: actions/setup-ruby@v1
      with:
        ruby-version: 2.6.x

    - name: Install PostgreSQL 11 client
      run: |
        sudo apt-get -yqq install libpq-dev

    - name: Build App
      env:
        PGHOST: localhost
        PGUSER: postgres
        RAILS_ENV: test
      run: |
        gem install bundler
        bundle install --jobs 4 --retry 3
        bin/rails db:setup

    - name: Run Tests
      env:
        PGHOST: localhost
        PGUSER: postgres
        RAILS_ENV: test
      run: |
      bundle exec rake test
      bundle exec rake test:system
      # or Rspec
      # bundle exec rspec
```

If you use SQL dumps—you have a `db/structure.sql` file rather than a `db/schema.rb` file in your application—you’ll need to use [this slightly altered workflow](github-actions-ci-for-rails-with-postgresql-11-and-structure-sql).

## Why?

> There is a benefit to your tests being in the ‘same place’ as your code. There is no separate account to maintain, no separate service to bill, and no separate team member administration.

Reviewer: Yesss!

> GitHub completely free for public, open source, repositories. There’s also a pretty generous free tier for private repositories, and it is pretty economical after that.

Reviewer: *GitHub Actions is
```suggestion
GitHub Actions is completely free for public, open source, repositories. There’s also a pretty generous free tier for private repositories, and it is pretty economical after that.
```

It can also be configured to do many tasks other than just running your tests: automatic image compression, other code checks, notifications to other systems, adding comments, making assignments, doing deployments, and more. All you have to do is fight with the configuration. :-)

## Why not?

Your testing infrastructure should, as much as possible, reflect your production environment to help catch subtle bugs. For me, at work for [CoverageBook](https://coveragebook.com) & [AnswerThePublic](https://answerthepublic.com), that means using Heroku’s [pipelines, CI, and review apps](https://www.heroku.com/flow).

You might think that increased centralisation of your development infrastructure at one company is a risk, or that a different CI provider might offer higher levels of service or configurability if you have specific needs.

> If you have a testing infrastructure that’s working well somewhere else, it probably isn’t worth the effort to move, although you could set up a parallel CI on GitHub to compare with your existing provider.

Reviewer: *compare against
```suggestion
If you have a testing infrastructure that’s working well somewhere else, it probably isn’t worth the effort to move, although you could set up a parallel CI on GitHub to compare against your existing provider.
```

---
title: "Use GitHub Actions for Rails CI with Postgres"
description: "So many commits to get here"
layout: article
category: ruby
image:
  base: '2019/github-actions-ci-for-rails-with-postgresql'
  alt: 'Green Octocat testing'

---

> Writing tests alongside writing software is strongly recommended. It helps you protect against bugs, reduce the fear of scary deploys and, can even help you write better software.

Reviewer: ```suggestion
Writing tests alongside your software is strongly recommended. It helps you protect against bugs, reduces the fear of scary deploys, and can even help you write better software.
```

The Ruby and Rails community have always had a strong focus on testing. Typically that means having some sort of “continuous integration” system constantly running your test suite.

> There’s a bunch of good choices, but recently GitHub launched [Actions](https://github.com/features/actions) (in beta), which lets you run arbitrary workflows, including tests, after certain things happen to your git repository.

Reviewer: *within (?) your git repository

> It does need a little extra configuration to run your tests using PostgreSQL (my SQL database of choice), if you use `structure.sql` database dumps [you’ll need a slightly different configuration](/ruby/github-actions-ci-for-rails-with-postgresql-11-and-structure-sql).

Reviewer: No comma  after (my SQL db of choice)

## Instead of...

> ...using a separate CI service.

Reviewer: comma instead of full stop

## Use...

...Github Actions to run your Rails tests.

### `.github/workflows/tests.yml`

```yml
name: Rails Tests

on:
  pull_request:
    - 'master'
  push:
    - 'master'

jobs:
  build:
    runs-on: ubuntu-latest

    services:
      postgres:
        image: postgres:11.5
        ports: ["5432:5432"]
        options: --health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 5

    steps:
    - uses: actions/checkout@v1

    - name: Set up Ruby 2.6
      uses: actions/setup-ruby@v1
      with:
        ruby-version: 2.6.x

    - name: Install PostgreSQL 11 client
      run: |
        sudo apt-get -yqq install libpq-dev

    - name: Build App
      env:
        PGHOST: localhost
        PGUSER: postgres
        RAILS_ENV: test
      run: |
        gem install bundler
        bundle install --jobs 4 --retry 3
        bin/rails db:setup

    - name: Run Tests
      env:
        PGHOST: localhost
        PGUSER: postgres
        RAILS_ENV: test
      run: |
      bundle exec rake test
      bundle exec rake test:system
      # or Rspec
      # bundle exec rspec
```

> If you use SQL dumps—you have a `db/structure.sql` file rather than a `db/schema.rb` file in your application—you’ll need to use [this slightly altered workflow](github-actions-ci-for-rails-with-postgresql-11-and-structure-sql).

Reviewer: Maybe get rid of the time you mention this above then?

## Why?

> There is a benefit to your tests being in the ‘same place’ as your code. There is no separate account to maintain, no separate service to bill, and no separate ‘team member management’.

Reviewer: Why inverted commas? Are they really needed? They are in the same place! :-D

Reviewer: *'team member management' to administer/worry about (?)

> It's completely free for public, open source, repositories. There’s also a pretty generous free usage tier for private repositories, and pretty economical after that.

Reviewer: *GitHub Actions is completely free... (since we haven't said the name in a while)

Reviewer: *free-usage tier or just free tier?

Reviewer: *, and it's pretty economical beyond that.

> Github Actions can also be configured to do many tasks other than just running your tests: automatic image compression, other code checks, notifications to other systems, adding comments, making assignments, doing deployments, and more. All you have to do is fight with the configuration. :-)

Reviewer: *It can also be configured (if we start the paragraph before with 'GitHub Actions')

## Why not?

> Your testing infrastructure should, as much as possible, reflect your production environment to help catch subtle bugs. For me, at work for [CoverageBook](https://coveragebook.com) & [AnswerThePublic](https://answerthepublic.com), that means using Heroku’s [pipelines, CI and review apps](https://www.heroku.com/flow).

Reviewer: *CI, and review apps (yes, that's an extra comma. :P)

Reviewer: Feel free to ignore it, if you must...

You might think that increased centralisation of your development infrastructure at one company is a risk, or that a different CI provider might offer higher levels of service or configurability if you have specific needs.

If you have a testing infrastructure that’s working well somewhere else, it probably isn’t worth the effort to move, although you could set up a parallel CI on GitHub to compare with your existing provider.

### Specific Ruby versions

> There’s an current issue with with the provided versions of Ruby. GitHub only provides certain versions [in their architecture](https://help.github.com/en/github/automating-your-workflow-with-github-actions/software-in-virtual-environments-for-github-actions#ubuntu-1804-lts) (scroll down for compatible versions) and it seems, from [this comment](https://github.com/actions/setup-ruby/issues/14#issuecomment-524020179), that releasing new CI images when Ruby is updated isn’t a priority.

Reviewer: *_a_ current issue...
